File: examples/typescript/src/core/logger.ts
// src/core/logger.ts — Structured logging utility

export interface LogEntry {
  level: "debug" | "info" | "warn" | "error";
  message: string;
  context: Record<string, unknown>;
  timestamp: Date;
}

export interface Logger {
  debug(message: string, context?: Record<string, unknown>): void;
  info(message: string, context?: Record<string, unknown>): void;
  warn(message: string, context?: Record<string, unknown>): void;
  error(message: string, context?: Record<string, unknown>): void;
}

export class ConsoleLogger implements Logger {
  debug(message: string, context?: Record<string, unknown>): void {
    this.write("debug", message, context);
  }
  info(message: string, context?: Record<string, unknown>): void {
    this.write("info", message, context);
  }
  warn(message: string, context?: Record<string, unknown>): void {
    this.write("warn", message, context);
  }
  error(message: string, context?: Record<string, unknown>): void {
    this.write("error", message, context);
  }

  private write(level: string, message: string, context?: Record<string, unknown>): void {
    const entry: LogEntry = {
      level: level as LogEntry["level"],
      message,
      context: context || {},
      timestamp: new Date(),
    };
    const line = JSON.stringify(entry);
    if (entry.level === "error") {
      console.error(line);
    } else if (entry.level === "warn") {
      console.warn(line);
    } else {
      console.log(line);
    }
  }
}

File: examples/dart/lib/core/logger.dart
import 'dart:convert';
import 'dart:io';

class Logger {
  void debug(String message, {Map<String, dynamic>? context}) {
    _log('debug', message, context);
  }

  void info(String message, {Map<String, dynamic>? context}) {
    _log('info', message, context);
  }

  void warn(String message, {Map<String, dynamic>? context}) {
    _log('warn', message, context);
  }

  void error(String message, {Map<String, dynamic>? context}) {
    _log('error', message, context);
  }

  void _log(String level, String message, Map<String, dynamic>? context) {
    final map = <String, dynamic>{
      'level': level,
      'message': message,
      'timestamp': DateTime.now().toIso8601String(),
    };
    if (context != null) map['context'] = context;
    final line = const JsonEncoder.withIndent('  ').convert(map);
    if (level == 'error' || level == 'warn') {
      stderr.writeln(line);
    } else {
      stdout.writeln(line);
    }
  }
}
